File: Older Versions/V2.2/RAM.py
def write(id: int, new_data: int, signed: bool = True):
    with open("RAM.bin", "rb") as file:
        data = bytearray(file.read())
        if (
            signed and -128 <= new_data < 0
        ):  # converts the negative number to unsigned number
            data[id] = 256 + new_data
        elif (not signed and 0 <= new_data <= 255) or (
            signed and 0 <= new_data <= 127
        ):  # sb doesnt matter, write number
            data[id] = new_data
        else:
            raise OverflowError(f"Cannot write {new_data} to register {id}")

    with open("RAM.bin", "wb") as file:
        file.write(data)

def read(id: int, signed: bool = True):
    with open("RAM.bin", "rb") as f:
        value = f.read()[id]
    if not signed:  # 0 to 255
        return value
        # -128 to 127
    if value > 127:  # when sig bit is 1, meaning negative
        return -(256 - value)
    else:  # when sig bit is 0, meaning positive
        return value

File: Excpute-JS/RAM.js
var RAM = Array(256).fill(0);

function ram_read(id, signed_value = true) {
  let value = RAM[id];
  if (!signed_value) {
    return value;
  }
  if (value > 127) {
    return -(256 - value);
  } else {
    return value;
  }
}

function ram_write(id, new_data, signed_value = true) {
  if (signed_value && -128 <= new_data && new_data < 0) {
    RAM[id] = 256 + new_data;
  } else if ((!signed_value && 0 <= new_data && new_data <= 255) || (signed_value && 0 <= new_data && new_data <= 127)) {
    RAM[id] = new_data;
  } else {
    throw new Error(`Cannot write ${new_data} to register ${id}`);
  }
}

module.exports = { ram_read, ram_write };
